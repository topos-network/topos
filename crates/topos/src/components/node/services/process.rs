use crate::edge::CommandConfig;
use std::collections::HashMap;
use std::path::PathBuf;
use std::process::ExitStatus;
use thiserror::Error;
use tokio::{spawn, sync::mpsc, task::JoinHandle};
use tokio_util::sync::CancellationToken;
use topos_config::sequencer::SequencerConfig;
use topos_config::tce::{AuthKey, StorageConfiguration, TceConfig};
use topos_p2p::Multiaddr;
use topos_sequencer::SequencerConfiguration;
use topos_tce_transport::ReliableBroadcastParams;
use topos_wallet::SecretManager;
use tracing::{debug, error, info, warn};

use topos_config::genesis::Genesis;

#[derive(Error, Debug)]
pub enum Errors {
    #[error("TCE error")]
    TceFailure,
    #[error("Sequencer error")]
    SequencerFailure,
    #[error("Edge error: {0}")]
    EdgeTerminated(#[from] std::io::Error),
}

pub fn generate_edge_config(
    edge_path: PathBuf,
    config_path: PathBuf,
) -> JoinHandle<Result<ExitStatus, Errors>> {
    // Create the Polygon Edge config
    info!("Generating the configuration at {config_path:?}");
    info!("Polygon-edge binary located at: {edge_path:?}");
    spawn(async move {
        CommandConfig::new(edge_path)
            .init(&config_path)
            .spawn()
            .await
            .map_err(|e| {
                error!("Failed to generate the edge configuration: {e:?}");
                Errors::EdgeTerminated(e)
            })
    })
}

pub(crate) fn spawn_sequencer_process(
    config: SequencerConfig,
    keys: &SecretManager,
    shutdown: (CancellationToken, mpsc::Sender<()>),
) -> JoinHandle<Result<ExitStatus, Errors>> {
    let config = SequencerConfiguration {
        subnet_id: config.subnet_id,
        public_key: keys.validator_pubkey(),
        subnet_jsonrpc_http: config.subnet_jsonrpc_http,
        subnet_jsonrpc_ws: config.subnet_jsonrpc_ws,
        subnet_contract_address: config.subnet_contract_address,
        tce_grpc_endpoint: config.tce_grpc_endpoint,
        signing_key: keys.validator.clone().unwrap(),
        verifier: 0,
        start_block: config.start_block,
    };

    debug!("Sequencer args: {config:?}");
    spawn(async move {
        topos_sequencer::run(config, shutdown).await.map_err(|e| {
            error!("Sequencer failure: {e:?}");
            Errors::SequencerFailure
        })
    })
}

pub(crate) fn spawn_tce_process(
    mut config: TceConfig,
    keys: SecretManager,
    genesis: Genesis,
    shutdown: (CancellationToken, mpsc::Sender<()>),
) -> JoinHandle<Result<ExitStatus, Errors>> {
    config.boot_peers = genesis
        .boot_peers(Some(topos_p2p::constants::TCE_BOOTNODE_PORT))
        .into_iter()
        .chain(config.parse_boot_peers())
        .collect::<Vec<_>>();
    config.auth_key = keys.network.map(AuthKey::PrivateKey);
    config.signing_key = keys.validator.map(AuthKey::PrivateKey);
    config.p2p.is_bootnode = if let Some(AuthKey::PrivateKey(ref k)) = config.auth_key {
        let peer_id = topos_p2p::utils::keypair_from_protobuf_encoding(&k[..])
            .public()
            .to_peer_id();

        config.boot_peers.iter().any(|(p, _)| p == &peer_id)
    } else {
        false
    };

    config.validators = genesis.validators().expect("Cannot parse validators");
    config.tce_params = ReliableBroadcastParams::new(config.validators.len());

    if let Some(socket) = config.libp2p_api_addr {
        warn!(
            "`libp2p_api_addr` is deprecated in favor of `listen_addresses` and \
             `public_addresses` and will be removed in the next version. In order to keep your \
             node running, `libp2p_api_addr` will be used."
        );

        let addr: Multiaddr = format!("/ip4/{}/tcp/{}", socket.ip(), socket.port())
            .parse()
            .expect("Unable to generate Multiaddr from `libp2p_api_addr`");

        config.p2p.listen_addresses = vec![addr.clone()];
        config.p2p.public_addresses = vec![addr];
    }

    config.version = env!("TOPOS_VERSION");
    config.storage = StorageConfiguration::RocksDB(Some(config.db_path.clone()));

    debug!("TCE args: {config:?}");
    spawn(async move {
        topos_tce::run(&config, shutdown).await.map_err(|e| {
            error!("TCE process terminated: {e:?}");
            Errors::TceFailure
        })
    })
}

pub fn spawn_edge_process(
    edge_path: PathBuf,
    data_dir: PathBuf,
    genesis_path: PathBuf,
    edge_args: HashMap<String, String>,
) -> JoinHandle<Result<ExitStatus, Errors>> {
    spawn(async move {
        CommandConfig::new(edge_path)
            .server(&data_dir, &genesis_path, edge_args)
            .spawn()
            .await
            .map_err(Errors::EdgeTerminated)
    })
}
