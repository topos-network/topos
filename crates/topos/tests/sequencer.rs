mod utils;

use std::process::Command;

use assert_cmd::prelude::*;

#[test]
fn sequencer_help_display() -> Result<(), Box<dyn std::error::Error>> {
    let mut cmd = Command::cargo_bin("topos")?;
    cmd.arg("sequencer").arg("run").arg("-h");

    let output = cmd.assert().success();

    let result: &str = std::str::from_utf8(&output.get_output().stdout)?;

    insta::assert_snapshot!(utils::sanitize_config_folder_path(result));

    Ok(())
}
