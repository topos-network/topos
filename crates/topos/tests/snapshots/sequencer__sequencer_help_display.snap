---
source: crates/topos/tests/sequencer.rs
expression: result
---
Run a full Topos Sequencer instance

Usage: topos sequencer run [OPTIONS] --subnet-contract-address <SUBNET_CONTRACT_ADDRESS> --subnet-data-dir <SUBNET_DATA_DIR>

Options:
      --subnet-id <SUBNET_ID>
          SubnetId of the local subnet node, hex encoded 32 bytes starting with 0x [env: TOPOS_LOCAL_SUBNET_ID=]
  -v, --verbose...
          Defines the verbosity level
      --home <HOME>
          Home directory for the configuration [env: TOPOS_HOME=] [default: /home/runner/.config/topos]
      --subnet-jsonrpc-endpoint <SUBNET_JSONRPC_ENDPOINT>
          [env: SUBNET_JSONRPC_ENDPOINT=] [default: 127.0.0.1:8545]
      --subnet-contract-address <SUBNET_CONTRACT_ADDRESS>
          [env: SUBNET_CONTRACT_ADDRESS=]
      --base-tce-api-url <BASE_TCE_API_URL>
          Base Uri of TCE node to call grpc service api [env: TOPOS_BASE_TCE_API_URL=] [default: http://[::1]:1340]
      --subnet-data-dir <SUBNET_DATA_DIR>
          Polygon subnet node data dir, containing `consensus/validator.key`, e.g. `../test-chain-1` [env: TOPOS_LOCAL_SUBNET_DATA_DIR=]
      --verifier <VERIFIER>
          Verifier version [env: TOPOS_SEQUENCER_VERIFIER_VERSION=] [default: 0]
      --otlp-agent <OTLP_AGENT>
          Socket of the opentelemetry agent endpoint If not provided open telemetry will not be used [env: TOPOS_OTLP_AGENT=]
      --otlp-service-name <OTLP_SERVICE_NAME>
          Otlp service name If not provided open telemetry will not be used [env: TOPOS_OTLP_SERVICE_NAME=]
  -h, --help
          Print help

